File: bench/compare/build_bench.py
"""Time JAX `gp.build_graph` on CPU (steady-state; excludes the one-time JIT compile).

    JAX_PLATFORMS=cpu python build_bench.py [N] [K] [D]

Prints one `BUILD {...}` JSON line. Pair with build_bench.jl (run_build_compare.sh runs both
under matched cores). Needs a Python env with jax + graphgp.
"""
import sys
import time
import json

import numpy as np
import jax
import jax.numpy as jnp
import graphgp as gp

N = int(sys.argv[1]) if len(sys.argv) > 1 else 1_000_000
K = int(sys.argv[2]) if len(sys.argv) > 2 else 10
D = int(sys.argv[3]) if len(sys.argv) > 3 else 3

pts = jnp.asarray(np.random.RandomState(1).randn(N, D).astype(np.float32))
g = gp.build_graph(pts, n0=256, k=K)
jax.block_until_ready(g.points)                          # warm + compile
t0 = time.perf_counter()
g = gp.build_graph(pts, n0=256, k=K)
jax.block_until_ready(g.points)
dt = time.perf_counter() - t0

print("BUILD " + json.dumps({"impl": "jax-cpu", "N": N, "K": K, "D": D,
                             "platform": jax.devices()[0].platform, "seconds": round(dt, 3)}))

File: bench/compare/build_bench.jl
# Time GraphGP.jl `build_graph` (parallel CPU). Mirrors build_bench.py.
#
#   julia -t <N> --project=bench bench/compare/build_bench.jl [N] [K] [D]
#
# Prints one `BUILD {...}` JSON line. run_build_compare.sh runs this and build_bench.py under
# matched cores for the head-to-head.

using GraphGP
using Random
using Printf

N = length(ARGS) >= 1 ? parse(Int, ARGS[1]) : 1_000_000
K = length(ARGS) >= 2 ? parse(Int, ARGS[2]) : 10
D = length(ARGS) >= 3 ? parse(Int, ARGS[3]) : 3

bins, vals = rbf_kernel(1.0, 0.3, 1e-4, 1e1, 200; jitter = 1e-2)
build_graph(randn(500, D), 50, 8, bins, vals)            # warm (compile)
P = randn(MersenneTwister(1), N, D)
t = @elapsed build_graph(P, 256, K, bins, vals)

@printf("BUILD {\"impl\":\"graphgp.jl-cpu\",\"N\":%d,\"K\":%d,\"D\":%d,\"threads\":%d,\"seconds\":%.3f}\n",
    N, K, D, Threads.nthreads(), t)
